// @statoscope/webpack-model/dist/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nodeModule = exports.moduleReasonResource = exports.moduleResource = exports.moduleNameResource = exports.matchRxValue = exports.concatenatedIdRx = exports.extractFileRx = void 0;
exports.extractFileRx = /!?([^!]+)$/;
exports.concatenatedIdRx = /(.+) \+ \d+ modules$/;
function matchRxValue(rx, string) {
    const [, match] = string.match(rx) || [];
    return match || null;
}
exports.matchRxValue = matchRxValue;
const resourceNameCache = new Map();
const nodeModuleNameCache = new Map();
function moduleNameResource(name) {
    if (name) {
        const cached = resourceNameCache.get(name);
        if (cached) {
            return cached;
        }
        if (!name.includes('(ignored)') && !name.startsWith('multi')) {
            const normalized = matchRxValue(exports.extractFileRx, name.replace('(webpack)', 'node_modules/webpack'));
            if (!normalized) {
                resourceNameCache.set(name, name);
                return name;
            }
            const nameResource = matchRxValue(exports.concatenatedIdRx, normalized) || normalized;
            if (nameResource.startsWith('./') || nameResource.startsWith('.\\')) {
                const result = nameResource.slice(2);
                resourceNameCache.set(name, result);
                return result;
            }
            resourceNameCache.set(name, nameResource);
            return nameResource;
        }
    }
    return null;
}
exports.moduleNameResource = moduleNameResource;
function moduleResource(module) {
    var _a;
    if ((module === null || module === void 0 ? void 0 : module.moduleType) &&
        ((module === null || module === void 0 ? void 0 : module.moduleType) === 'provide-module' ||
            (module === null || module === void 0 ? void 0 : module.moduleType) === 'consume-shared-module')) {
        return module.name;
    }
    return moduleNameResource((_a = module === null || module === void 0 ? void 0 : module.name) !== null && _a !== void 0 ? _a : null);
}
exports.moduleResource = moduleResource;
function moduleReasonResource(reason) {
    var _a;
    return moduleNameResource((_a = reason === null || reason === void 0 ? void 0 : reason.moduleName) !== null && _a !== void 0 ? _a : null);
}
exports.moduleReasonResource = moduleReasonResource;
function nodeModule(path) {
    if (!path) {
        return null;
    }
    let cached = nodeModuleNameCache.get(path);
    if (!cached) {
        const lastNodeModulesRx = /.*node_modules[/\\](?:(@.+?)[/\\])?([^/\\]+)/;
        const [input, namespace, name] = path.match(lastNodeModulesRx) || [];
        const isRoot = input
            ? input.indexOf('node_modules') === input.lastIndexOf('node_modules')
            : false;
        cached = name
            ? {
                path: input,
                name: [namespace, name].filter(Boolean).join('/'),
                isRoot,
            }
            : null;
        nodeModuleNameCache.set(path, cached);
    }
    return cached;
}
exports.nodeModule = nodeModule;
